File: app/src/shared/domain/schedulingKind.ts
import type { RecurrenceEndInput } from './corePort';

export type SchedulingKind = 'one_shot' | 'recurring';

type SchedulingKindCarrier = {
  origin?: 'one_shot' | 'recurring';
  scheduleKind?: 'one_shot' | 'recurring';
  recurrenceEnd?: RecurrenceEndInput;
};

export function resolveSchedulingKind(input: SchedulingKindCarrier): SchedulingKind {
  if (input.origin === 'one_shot' || input.scheduleKind === 'one_shot') {
    return 'one_shot';
  }
  if (input.origin === 'recurring' || input.scheduleKind === 'recurring') {
    return 'recurring';
  }
  if (input.recurrenceEnd?.kind === 'after_occurrences' && input.recurrenceEnd.afterOccurrences === 1) {
    return 'one_shot';
  }
  return 'recurring';
}

File: core/src/main/kotlin/com/gonezo/domain/recurrence/SchedulingKind.kt
package com.gonezo.recurrence.domain

enum class SchedulingKind(val value: String) {
    RECURRING("recurring"),
    ONE_SHOT("one_shot"),
    ;

    companion object {
        fun from(value: String): SchedulingKind = entries.firstOrNull { it.value == value.trim().lowercase() }
            ?: throw IllegalArgumentException("Unsupported scheduling kind: $value")
    }
}

fun resolveCurrentSchedulingKind(recurrenceEnd: RecurrenceEnd): SchedulingKind = if (recurrenceEnd is RecurrenceEnd.AfterOccurrences && recurrenceEnd.count == 1) {
    SchedulingKind.ONE_SHOT
} else {
    SchedulingKind.RECURRING
}
